tools/xenpcid/xenpcid.c

```diff
@@ -0,0 +1,675 @@
+                                                    enum pcid__json_node_type expected_type)
+{
+    struct flexarray *maps = NULL;
+    int idx = 0;
+    struct pcid__json_map_node *node = NULL;
+
+    if (pcid__json_object_is_map(o)) {
+        maps = o->u.map;
+        for (idx = 0; idx < maps->count; idx++) {
+            if (flexarray_get(maps, idx, (void**)&node) != 0) {
+                return NULL;
+            }
+
+            if (strcmp(key, node->map_key) == 0) {
+                if (expected_type == JSON_ANY
+                    || (node->obj && (node->obj->type & expected_type))) {
+                    return node->obj;
+                } else
+                    return NULL;
+            }
+        }
+    }
+    return NULL;
+}
+
+static inline bool pcid__json_object_is_array(const struct pcid__json_object *o)
+{
+    return o != NULL && o->type == JSON_ARRAY;
+}
+
+static int pcid__json_object_append_to(struct pcid__json_object *obj,
+                                       struct pcid__yajl_ctx *ctx_yajl)
+{
+    struct pcid__json_object *dst = ctx_yajl->current;
+
+    if (dst) {
+        switch (dst->type) {
+        case JSON_MAP: {
+            struct pcid__json_map_node *last = NULL;
+
+            if (dst->u.map->count == 0) {
+                fprintf(stderr,
+                        "Try to add a value to an empty map (with no key)\n");
+                return ERROR_FAIL;
+            }
+            flexarray_get(dst->u.map, dst->u.map->count - 1, (void**)&last);
+            last->obj = obj;
+            break;
+        }
+        case JSON_ARRAY:
+            flexarray_append(dst->u.array, obj);
+            break;
+        default:
+            fprintf(stderr,
+                    "Try append an object is not a map/array (%i)\n",
+                    dst->type);
+            return ERROR_FAIL;
+        }
+    }
+
+    obj->parent = dst;
+
+    if (pcid__json_object_is_map(obj) || pcid__json_object_is_array(obj))
+        ctx_yajl->current = obj;
+    if (ctx_yajl->head == NULL)
+        ctx_yajl->head = obj;
+
+    return 0;
+}
+
+static int json_callback_string(void *opaque, const unsigned char *str,
+                                long unsigned int len)
+{
+    struct pcid__yajl_ctx *ctx = opaque;
+    char *t = NULL;
+    struct pcid__json_object *obj = NULL;
+
+    t = pcid_zalloc(len + 1);
+    strncpy(t, (const char *) str, len);
+    t[len] = 0;
+
+    obj = pcid_zalloc(sizeof(*obj));
+    obj->u.string = t;
+
+    if (pcid__json_object_append_to(obj, ctx))
+        return 0;
+
+    return 1;
+}
+
+static struct pcid__json_object *pcid__json_object_alloc(enum pcid__json_node_type type)
+{
+    struct pcid__json_object *obj;
+
+    obj = pcid_zalloc(sizeof(*obj));
+    obj->type = type;
+    obj->u.map = NULL;
+    obj->u.array = NULL;
+    obj->u.string = NULL;
+
+    if (type == JSON_MAP || type == JSON_ARRAY) {
+        struct flexarray *array = flexarray_make(1, 1);
```

Comment:
Missing empty line between variable declaration and code

Reply:
Done